### ISSUE TITLE
cobyla returns last evaluated function which might not be minimum

### ISSUE THREAD

**Issue by ericjster** [stevengj/nlopt#57]
COBYLA returns the wrong value for minimum.
It returns the last evaluated function value, but that last value might not be the return value.
Attached is an example of minimizing a test function ("Powell's Badly Scaled Function").
NEWUOA and BOBYQA do well.

The output shows the minima is found (close to 0.0), but then some additional evaluations are done and it seems to forget the best value.
...
myfunc: x: [  5.39347657e+00   2.17974583e-05] , val: 0.0308692676819
myfunc: x: [  5.39307750e+00  -2.79444985e-04] , val: 258.266926113
myfunc: x: [  5.39385142e+00  -3.09088486e-04] , val: 312.291607364
optimum at  5.39385141967 -0.000309088485691
minimum value =  312.291607364
result code =  4

[nlopt_cobyla_bug.py.txt](https://github.com/stevengj/nlopt/files/120082/nlopt_cobyla_bug.py.txt)

**Comment by jschueller** [comment]
confirmed, this is a bug in cobyla.c, appears in the original powell version as well.

**Comment by jschueller** [comment]
confirmed, this is a bug in cobyla.c, appears in the original powell version as well.

@stevengj, maybe the algorithm does not check for optimality for each feasible point of the simplex after it's evaluation with calcfc (it does not go to L620 but "loops"). Maybe something could be added as you did around line 600 ? The code is quite hard to follow because of these gotos, I hope you still know the code : )

**Comment by E-B41** [comment]
I think we have problems with the same bug.
The algorithm does not always return the optimum set of variables + these variables might violate given constraints. Looking at the course of optimization variables there are sets, that are okay.
This seems also be related to Issue #182 .

Is there any progress on this bug?
@jschueller is it correct, that the OpenTURNS implementation saves the optimization progress and uses it to return the correct result?

**Comment by markusbrugger** [comment]
We have stumbled across this bug when finding the minimum of a ratio of polynomials with several inequality constraints using nlopt's COBYLA algorithm. We are quite happy with the nlopt library, because of its neat and plain integration as a C library into our C program. However, the above-mentioned bug is still present in the current version 2.6.2. Our abject question hence is, why there has been no progress on this issue during the past? Surely, a work-around is to keep track of the optimization history manually (what we actually do now) and check in the end if COBYLA returns the correct values, otherwise take the really correct values from the external array that keeps the best result of the objective function together with the corresponding, non-violating parameter set. This procedure seems and is in fact clumsy, especially considering the fact that many users might not be aware of the bug, because their final values can rarely be called "suspicious". In our case, the bug was straightforward to detect, because before COBYLA takes over the steering wheel, a first round of a manual optimization takes place, whose optimized parameters are then handed over to COBYLA as valid starting values. In the end, COBYLA from time to time ends up with a score of, say, -20.0, although it started with, say, 2.5 (the larger the better in this case). We think, fixing this bug should not be too demanding (see also jschueller's comment from 2016 above), but it would prevent the scientific world from a bunch of false-positive/negative results due to a tiny, but nasty bug in the return behaviour of COBYLA.
We are sure that fixing this bug would greatly improve the nlopt library and substantially contributes to its trustworthyness.

We cordially hope for some comments on this topic and would also offer help to find a solution to the problem!

**Comment by matthewelmer-tamu** [comment]
This seems like a bug worth mentioning at the very top of the README...

**Comment by matthewelmer-tamu** [comment]
Actually, seeing as this absolutely catastrophic bug is still in the code after eight (8) years, I'd say it's high time to archive the repo @stevengj.
